src/common/commonutils/FileUtils.c

```diff
@@ -120,25 +120,52 @@ bool SavePayloadToFile(const char* fileName, const char* payload, const int payl
 
 bool AppendPayloadToFile(const char* fileName, const char* payload, const int payloadSizeBytes, void* log)
 {
-    char* fileContents = NULL;
+    FILE* existingFile = NULL;
+    struct stat st;
     bool result = false;
+    char lastCharacter = EOF;
+
+    if(NULL == fileName || NULL == payload || 0 == payloadSizeBytes)
+    {
+        OsConfigLogError(log, "AppendPayloadToFile: invalid arguments");
+        return false;
+    }
 
-    // If the file exists and there is no EOL at the end of file, add one before the append
-    if ((NULL != payload) && (payloadSizeBytes > 0) && FileExists(fileName) && 
-        (NULL != (fileContents = LoadStringFromFile(fileName, false, log))) && 
-        (EOL != fileContents[strlen(fileContents) - 1]))

```

Comment:
What exactly was wrong with this approach? Please elaborate. This function is used for many ASB checks which have been fully validated. Fixing one security bug found by fuzzing should be as minimal as possible. Do not refactor an entire function without strong reason and state what that is.

Reply:
> Unclear here what exactly was wrong with this approach? Can you elaborate?
There are couple of bugs and issues in this single statement:
- we load the whole file into memory just to read one byte,
- we don't check whether length of the fileContents is actually greater than 0,
- we perform a complex operation like allocating memory to fileContents and we have no other choice than freeing it in different place, because it's bundled together,
- why do we check for file existence first? Is the LoadStringFromFile gonna crash on non-existent file or it will just return NULL? We have a NULL check already,
- why don't we actually do a NULL check for the payload and payload size in the very beginning of the function? Is a NULL or empty payload a valid input to the function? If providing a NULL payload is invalid, in this construct we have no way of properly logging it,
- by bundling together all these operations code becomes much less clear. For example, one needs to look into the LoadStringFromFile to check if it returns a non-empty string. Lack of clarity is a very serious issue.

>  Fixing one security bug found by fuzzing should be as minimal as possible.

If the code is complex enough I think some minor upgrades should be done as we go, otherwise we'll only be adding complexity and there will never be good time to do refactoring.

> This function is used for many ASB checks which have been fully validated.

We should put into question the quality of validation then. Maybe we ask the validation team to actually do some code review and raise questions if something is unclear.

> We shall not refactor entire functions without strong reasons and then we need to state as detailed as possible what reasons are and why each part of the respective functions needs change.

I strongly believe we have reasons to do that.

Reply:
https://github.com/Azure/azure-osconfig/pull/790